scripts/build.sh

```diff
@@ -20,12 +20,33 @@ elif [[ "$OSTYPE" == "darwin"* ]]; then
     CMAKE_FLAGS+="-DCMAKE_C_COMPILER=${XCODE_CMDLINE_DIR}/usr/bin/clang -DCMAKE_CXX_COMPILER=${XCODE_CMDLINE_DIR}/usr/bin/clang++ -DCMAKE_CXX_FLAGS=-isystem\ /usr/local/include -DCMAKE_EXPORT_COMPILE_COMMANDS=ON"
 fi
 
-CMAKE_BUILD_TYPE="Debug"
-if [[ "$BUILD_RELEASE" == "1" ]]; then
+# Note:
+# CMAKE_BUILD_TYPE="Debug" adds "-O0 -g" flags by default
+# CMAKE_BUILD_TYPE="Release" adds "-O3 -DNDEBUG" by default
+if [[ "$BUILD_DEBUG" == "1" ]]; then
+    CMAKE_BUILD_TYPE="Debug"
+elif [[ "$BUILD_RELEASE" == "1" ]]; then
     CMAKE_BUILD_TYPE="Release"
 elif [[ "$BUILD_PROFILING" == "1" ]]; then
     CMAKE_BUILD_TYPE="Profiling"
 fi
 
+if [ $# -gt 0 ]; then
+    if [ "$1" == "Release" ]; then
+        CMAKE_BUILD_TYPE="Release"
+    elif [ "$1" == "Profiling" ]; then
+        CMAKE_BUILD_TYPE="Profiling"
+    elif [ "$1" == "Debug" ]; then
+        CMAKE_BUILD_TYPE="Debug"
+    fi
+fi
+
+if [[ -z $CMAKE_BUILD_TYPE ]]; then
+    echo "CMAKE_BUILD_TYPE not set, defaulting to debug"
+    CMAKE_BUILD_TYPE="Debug"
+fi
+
+echo "Building $CMAKE_BUILD_TYPE"
 eval "cmake -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE} ${CMAKE_FLAGS} .."
 make -j$CPUS
```

Comment:
[optional] We could print out error messages for debug if the cmake and make commands do not exit 0 as expected. 
This block could be put right after the `eval "cmake ...` line.
An identical copy of it could be after the `make ...` command except changing 'cmake' to 'make'.
```suggestion
if [[ $? -ne 0 ]]; then
    echo "cmake command failed"
    exit 1
fi
```

Reply:
This is an interesting point, I hadn't considered adding additional help statements here, as the cmake and make steps will give significant feedback as to why the compilation failed. Not sure that we necessarily need this here but can be persuaded.

Reply:
The suggested change seems to remove the actual build with `make` which is probably a mistake, but should not be committed.

Beyond that, testing to see if `cmake` failed rather than blindly running `make` probably isn't a bad idea.

Reply:
We set this script to exit if any command fails, so we could print out this info but it would require not using `set -e`